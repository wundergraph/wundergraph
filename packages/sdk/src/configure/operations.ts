export interface OperationsConfiguration {
	defaultConfig: BaseOperationConfiguration;
	queries: ConfigureQuery;
	mutations: ConfigureMutation;
	subscriptions: ConfigureSubscription;
	custom?: any;
}

export interface BaseOperationConfiguration {
	authentication: {
		required: boolean;
	};
}

export interface QueryConfiguration extends BaseOperationConfiguration {
	caching: {
		enable: boolean;
		public: boolean;
		maxAge: number;
		staleWhileRevalidate: number;
	};
	liveQuery: {
		enable: boolean;
		pollingIntervalSeconds: number;
	};
}

export interface MutationConfiguration extends BaseOperationConfiguration {}

export interface SubscriptionConfiguration extends BaseOperationConfiguration {}

type OperationConfiguration = QueryConfiguration | MutationConfiguration | SubscriptionConfiguration;

export interface MutationConfiguration extends BaseOperationConfiguration {}

export interface SubscriptionConfiguration extends BaseOperationConfiguration {}

export type ConfigureOperation = (config: OperationConfiguration) => OperationConfiguration;
export type ConfigureQuery = (config: BaseOperationConfiguration) => QueryConfiguration;
export type CustomizeQuery = (config: QueryConfiguration) => QueryConfiguration;
export type ConfigureMutation = (config: BaseOperationConfiguration) => MutationConfiguration;
export type CustomizeMutation = (config: MutationConfiguration) => MutationConfiguration;
export type ConfigureSubscription = (config: BaseOperationConfiguration) => SubscriptionConfiguration;
export type CustomizeSubscription = (config: SubscriptionConfiguration) => SubscriptionConfiguration;

export interface WunderGraphOperationsConfig<T extends OperationsConfiguration> {
	operations?: T;
}

export const configureWunderGraphOperations = <GeneratedOperationConfiguration extends OperationsConfiguration>(
	config: WunderGraphOperationsConfig<GeneratedOperationConfiguration>
) => config.operations;

export const disableAuth = <Configs extends QueryConfiguration | MutationConfiguration | SubscriptionConfiguration>(
	config: Configs
): Configs => {
	return {
		...config,
		authentication: {
			required: false,
		},
	};
};

export const enableAuth = <Configs extends QueryConfiguration | MutationConfiguration | SubscriptionConfiguration>(
	config: Configs
): Configs => {
	return {
		...config,
		authentication: {
			required: true,
		},
	};
};

export const enableCaching = (config: QueryConfiguration): QueryConfiguration => ({
	...config,
	caching: { ...config.caching, enable: true },
});
