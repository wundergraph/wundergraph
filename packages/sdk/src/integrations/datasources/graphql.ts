import type { GraphQLIntrospection } from '../../definition';
import { defineDatasource } from './define-datasource';

export interface GraphQLDatasourceOptions extends Omit<GraphQLIntrospection, 'apiNamespace'> {
	namespace?: string;
}

export const graphql = defineDatasource<GraphQLDatasourceOptions>((config) => {
	const { namespace, ...introspectionConfig } = config;
	return {
		name: 'graphql',
		hooks: {
			'config:setup': async (options) => {
				const { introspect } = await import('../../definition');
				options.addApi(
					introspect.graphql({
						apiNamespace: namespace,
						...introspectionConfig,
					})
				);
			},
		},
	};
});
