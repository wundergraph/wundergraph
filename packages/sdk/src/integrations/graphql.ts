import type { GraphQLIntrospection } from '../definition';
import { defineDatasource } from './define-datasource';

export interface GraphQLDatasourceOptions extends Omit<GraphQLIntrospection, 'apiNamespace'> {
	namespace: string;
}

export const graphql = defineDatasource<GraphQLDatasourceOptions>((config) => {
	return {
		name: 'graphql',
		hooks: {
			'config:setup': async (options) => {
				const { introspect } = await import('../definition');
				options.addApi(
					introspect.graphql({
						apiNamespace: config.namespace,
						...config,
					})
				);
			},
		},
	};
});
