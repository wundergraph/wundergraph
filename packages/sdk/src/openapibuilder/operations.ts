import { JSONSchema7Definition, JSONSchema7TypeName } from 'json-schema';

export interface JSONSchemaParameterPath {
	path: string[];
	required: boolean;
	type: JSONSchema7TypeName;
}

// TODO: Add "default" values
// path syntax follows https://github.com/tidwall/sjson#path-syntax
export function buildPath(
	path: string[],
	required: boolean,
	obj: JSONSchema7Definition,
	paths: JSONSchemaParameterPath[]
) {
	if (typeof obj === 'boolean') return;

	// those nodes are limited due to recursion
	if (!obj.type) return;

	if (obj.type === 'array') {
		if (typeof obj.items === 'boolean') return;
		if (!obj.items) return;

		let items = [];
		if (!Array.isArray(obj.items)) items.push(obj.items);
		items.forEach((obj, index) => buildPath([...path, index.toString()], false, obj, paths));
		return;
	}

	if (obj.properties) {
		Object.keys(obj.properties).forEach((key) => {
			if (obj.properties) {
				buildPath([...path, key], obj.required?.includes(key) || required, obj.properties[key], paths);
			}
		});
		return;
	}

	paths.push({
		path,
		required,
		type: obj.type as JSONSchema7TypeName,
	});

	paths.sort(function (a, b) {
		return a.path.length - b.path.length;
	});
}
