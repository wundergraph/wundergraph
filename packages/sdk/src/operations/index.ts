export { createOperationFactory } from './operations';
export { z } from 'zod';
export type { ExtractInput, ExtractResponse } from './operations';
export type {
	GraphQLError,
	GraphQLErrorLocation,
	ClientOperationErrors,
	TypeScriptOperationErrorCodes,
} from '../client/errors';
export { OperationError, AuthorizationError, InputValidationError } from '../client/errors';
