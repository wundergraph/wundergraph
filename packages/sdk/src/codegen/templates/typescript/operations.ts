import { doNotEditHeader, Template, TemplateOutputFile } from '../../index';
import { CodeGenerationConfig } from '../../../configure';
import { formatTypeScript } from './index';
import { OperationType } from '@wundergraph/protobuf';
import Handlebars from 'handlebars';
import { template } from './operations.template';

export class Operations implements Template {
	generate(generationConfig: CodeGenerationConfig): Promise<TemplateOutputFile[]> {
		const model: OperationsModel = {
			operations: generationConfig.config.application.Operations.map((op) => ({
				name: op.Name,
				isMutation: op.OperationType === OperationType.MUTATION,
				isQuery: op.OperationType === OperationType.QUERY,
				isSubscription: op.OperationType === OperationType.SUBSCRIPTION,
			})),
		};
		const tmpl = Handlebars.compile(template);
		const content = tmpl(model);
		return Promise.resolve([
			{
				path: 'wundergraph.operations.ts',
				content: formatTypeScript(content),
				header: doNotEditHeader,
			},
		]);
	}
}

interface OperationsModel {
	operations: {
		name: string;
		isQuery: boolean;
		isMutation: boolean;
		isSubscription: boolean;
	}[];
}
