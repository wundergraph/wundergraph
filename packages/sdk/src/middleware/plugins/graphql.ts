import { FastifyPluginAsync } from 'fastify';
import fp from 'fastify-plugin';
import { GraphQLSchema } from 'graphql';
import {
	getGraphQLParameters,
	processRequest,
	renderGraphiQL,
	sendResult,
	shouldRenderGraphiQL,
	ExecutionContext as HelixExecutionContext,
} from 'graphql-helix';
import { BaseContext } from '../server';

export interface GraphQLServerConfig {
	serverName: string;
	schema: GraphQLSchema | Promise<GraphQLSchema>;
	apiNamespace?: string;
	enableGraphQLEndpoint?: boolean;
	routeUrl: string;
	// use baseContext to pass global data to the graphql context
	baseContext?: Object | Promise<Object>;
	// use the contextFactory to create a context for each request
	// the callback needs to be called with your object
	contextFactory?: (callback: (ctx: Object) => void) => Promise<void>;
	// by default, root fields will be prefixed with the api namespace
	// use skipRenameRootFields to skip this behaviour, e.g. if a field is already prefixed
	skipRenameRootFields?: string[];
	// implement resolverFactory to create a custom resolver
	customResolverFactory?: (executionContext: HelixExecutionContext & ExecutionContext) => Promise<{}>;
}

interface ExecutionContext {
	wundergraph: BaseContext;
}

const FastifyGraphQLPlugin: FastifyPluginAsync<GraphQLServerConfig> = async (fastify, config) => {
	const schema = await config.schema;
	const baseContext = await config.baseContext;

	fastify.route({
		method: ['GET', 'POST'],
		url: config.routeUrl,
		async handler(req, res) {
			const request = {
				body: req.body,
				headers: req.headers,
				method: req.method,
				query: req.query,
			};

			const pluginLogger = req.ctx.log.child({ server: config.serverName, plugin: 'graphql' });

			if (config.enableGraphQLEndpoint && shouldRenderGraphiQL(request)) {
				res.type('text/html');
				res.send(
					renderGraphiQL({
						endpoint: config.routeUrl,
					})
				);
			} else {
				const request = {
					body: req.body,
					headers: req.headers,
					method: req.method,
					query: req.query,
				};
				const { operationName, query, variables } = getGraphQLParameters(request);

				if (config.contextFactory) {
					await config.contextFactory(async (ctx) => {
						const result = await processRequest<ExecutionContext>({
							operationName,
							query,
							variables,
							request,
							schema,
							// @ts-ignore
							rootValueFactory: config.customResolverFactory,
							contextFactory: (): ExecutionContext => ({
								...baseContext,
								...ctx,
								wundergraph: {
									...req.ctx,
									log: pluginLogger,
								},
							}),
						});

						await sendResult(result, res.raw);
						res.sent = true;
					});
					return;
				}

				const result = await processRequest<ExecutionContext>({
					operationName,
					query,
					variables,
					request,
					schema,
					// @ts-ignore
					rootValueFactory: config.customResolverFactory,
					contextFactory: (): ExecutionContext => ({
						...baseContext,
						wundergraph: {
							...req.ctx,
							log: pluginLogger,
						},
					}),
				});

				await sendResult(result, res.raw);

				// skip async/await handling because we already sent the result with the raw http socket
				res.sent = true;
			}
		},
	});
};

export default fp(FastifyGraphQLPlugin, '3.x');
