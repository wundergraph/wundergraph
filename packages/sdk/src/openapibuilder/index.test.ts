import { JSONSchema7 } from 'json-schema';

import { OperationExecutionEngine, OperationType } from '@wundergraph/protobuf';
import { GraphQLOperation } from '../graphql/operations';
import { OpenApiBuilder, isValidOpenApiSchemaName } from './index';

const emptySchema = {
	type: 'object',
	properties: {},
	additionalProperties: false,
};

const operations = [
	{
		Name: 'GetRootObj',
		PathName: 'GetRootObj',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: emptySchema,
		ResponseSchema: emptySchema,
	},
	{
		Name: 'GetRootObj2',
		PathName: 'GetRootObj2',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: emptySchema,
		ResponseSchema: emptySchema,
	},
	{
		Name: 'GetUser',
		PathName: 'users/get',
		OperationType: OperationType.QUERY,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: emptySchema,
		ResponseSchema: emptySchema,
	},
	{
		Name: 'SubscribeToUser',
		PathName: 'users/subscribe',
		OperationType: OperationType.SUBSCRIPTION,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		VariablesSchema: emptySchema,
		ResponseSchema: emptySchema,
	},
	{
		Name: 'PutUser',
		PathName: 'users/put',
		OperationType: OperationType.MUTATION,
		ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
		AuthenticationConfig: { required: true },
		VariablesSchema: emptySchema,
		ResponseSchema: emptySchema,
	},
] as GraphQLOperation[];

const build = (operations: GraphQLOperation[]) => {
	const builder = new OpenApiBuilder({
		title: 'WunderGraph',
		version: '0',
		baseURL: 'http://localhost:9991',
	});
	return builder.build(operations);
};

describe('OpenAPI builder', () => {
	test('no operations', async () => {
		const apiTitle = 'WunderGraph';
		const apiVersion = '1';
		const apiPublicUrl = 'http://localhost:9991';
		const builder = new OpenApiBuilder({
			title: apiTitle,
			version: apiVersion,
			baseURL: apiPublicUrl,
		});
		const result = builder.build([]);
		expect(result.openapi).toBe('3.1.0');
		expect(result.info.title).toBe(apiTitle);
		expect(result.info.version).toBe(apiVersion);
		expect(result.servers[0].url).toBe(`${apiPublicUrl}/operations`);
	});

	test('operation properties', async () => {
		const operations = [
			{
				Name: 'Query',
				PathName: 'QueryPath',
				OperationType: OperationType.QUERY,
				ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
				VariablesSchema: {
					type: 'object',
					properties: {
						input: {
							type: 'string',
						},
					},
					additionalProperties: false,
				},
				ResponseSchema: {
					type: 'object',
					properties: {
						output: {
							type: 'string',
						},
					},
					additionalProperties: false,
				},
			},
			{
				Name: 'Mutation',
				PathName: 'MutationPath',
				OperationType: OperationType.MUTATION,
				ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
				VariablesSchema: {
					type: 'object',
					properties: {
						input: {
							type: 'string',
						},
					},
					required: ['string'],
					additionalProperties: false,
				},
				ResponseSchema: {
					type: 'object',
					properties: {
						output: {
							type: 'string',
						},
					},
					additionalProperties: false,
				},
			},
		] as unknown as GraphQLOperation[];
		const result = build(operations);

		const querySpec = result.paths['/QueryPath'];
		expect(querySpec).toBeDefined();
		expect(querySpec.post).toBeUndefined();
		expect(querySpec.get?.operationId).toBe('Query');
		expect(querySpec.get?.parameters?.[0].name).toBe('input');
		expect(querySpec.get?.parameters?.[0].in).toBe('query');
		expect(querySpec.get?.parameters?.[0].required).toBe(false);
		expect(querySpec.get?.parameters?.[0].schema).toStrictEqual({ type: 'string' });
		const queryResponse = querySpec.get?.responses['200'];
		expect(queryResponse).toBeDefined();
		expect(queryResponse?.content['application/json'].schema).toBe(operations[0].ResponseSchema);

		const mutationSpec = result.paths['/MutationPath'];
		expect(mutationSpec).toBeDefined();
		expect(mutationSpec.get).toBeUndefined();
		expect(mutationSpec.post?.operationId).toBe('Mutation');
		expect(mutationSpec.post?.requestBody?.content['application/json'].schema).toBe(operations[1].VariablesSchema);
		const mutationResponse = mutationSpec.post?.responses['200'];
		expect(mutationResponse).toBeDefined();
		expect(mutationResponse?.content['application/json'].schema).toBe(operations[1].ResponseSchema);
	});

	test('skip internal operations', async () => {
		const operations = [
			{
				Name: 'Query',
				PathName: 'QueryPath',
				OperationType: OperationType.QUERY,
			},
			{
				Name: 'InternalQuery',
				PathName: 'InternalQueryPath',
				OperationType: OperationType.QUERY,
				Internal: true,
			},
		] as unknown as GraphQLOperation[];
		const result = build(operations);
		expect(Object.keys(result.paths).length).toBe(1);
		expect(result.paths['/QueryPath']).toBeDefined();
		expect(result.paths['/InternalQueryPath']).toBeUndefined();
	});

	test('annotate operations', async () => {
		const operations = [
			{
				Name: 'Query',
				PathName: 'QueryPath',
				OperationType: OperationType.QUERY,
			},
			{
				Name: 'Mutation',
				PathName: 'MutationPath',
				OperationType: OperationType.MUTATION,
				AuthenticationConfig: { required: true },
			},
			{
				Name: 'Subscription',
				PathName: 'SubscriptionPath',
				OperationType: OperationType.SUBSCRIPTION,
			},
		] as unknown as GraphQLOperation[];

		const result = build(operations);
		expect(Object.keys(result.paths).length).toBe(3);
		const query = result.paths['/QueryPath'].get;
		expect(query?.['x-wundergraph-operation-type']).toBe('query');
		expect(query?.['x-wundergraph-requires-authentication']).toBe(false);
		const mutation = result.paths['/MutationPath'].post;
		expect(mutation?.['x-wundergraph-operation-type']).toBe('mutation');
		expect(mutation?.['x-wundergraph-requires-authentication']).toBe(true);
		const subscription = result.paths['/SubscriptionPath'].get;
		expect(subscription?.['x-wundergraph-operation-type']).toBe('subscription');
		expect(subscription?.['x-wundergraph-requires-authentication']).toBe(false);
	});

	test('rewrite references and generate valid names inside anyOf', () => {
		const operations = [
			{
				Name: 'Weather',
				PathName: 'weather/get',
				OperationType: OperationType.QUERY,
				ExecutionEngine: OperationExecutionEngine.ENGINE_GRAPHQL,
				VariablesSchema: emptySchema,
				ResponseSchema: {
					type: 'object',
					properties: {
						weather: {
							anyOf: [
								{
									$ref: '#/definitions/{readonlysummary:{readonlytitle:string|null;readonlydescription:string|null;readonlyicon:string|null;}|null;}',
								},
								{
									type: 'null',
								},
							],
						},
					},
					required: ['weather'],
					definitions: {
						'{readonlysummary:{readonlytitle:string|null;readonlydescription:string|null;readonlyicon:string|null;}|null;}':
							{
								type: 'string',
							},
					},
				},
			},
		] as unknown as GraphQLOperation[];

		const result = build(operations);
		const operation = result.paths['/weather/get'].get;
		const responseSchema = operation?.responses?.['200']?.content?.['application/json']?.schema;
		const weather = responseSchema?.properties?.['weather'];
		const firstAnyOf = (weather as JSONSchema7).anyOf?.[0];
		const firstRef = (firstAnyOf as JSONSchema7).$ref;
		expect(firstRef).toMatch(/#\/components\/schemas\/.*/);

		const refName = firstRef?.substring('#/components/schemas/'.length);
		expect(refName).toBeDefined();
		expect(isValidOpenApiSchemaName(refName!)).toBeTruthy();
		expect(result.components?.schemas?.[refName!]).toBeDefined();
	});

	test('OpenAPI Builder', async () => {
		const builder = new OpenApiBuilder({
			title: 'WunderGraph',
			version: '0',
			baseURL: 'http://localhost:9991',
		});
		const result = builder.build(operations);

		expect(result).toMatchSnapshot();
	});
});
