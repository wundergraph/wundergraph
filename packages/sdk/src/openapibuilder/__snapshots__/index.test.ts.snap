// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI builder OpenAPI Builder 1`] = `
{
  "components": {
    "schemas": {
      "InvalidInputError": {
        "properties": {
          "errors": {
            "items": {
              "properties": {
                "invalidValue": {},
                "message": {
                  "type": "string",
                },
                "propertyPath": {
                  "type": "string",
                },
              },
              "required": [
                "propertyPath",
                "invalidValue",
                "message",
              ],
              "type": "object",
            },
            "type": "array",
          },
          "input": {},
          "message": {
            "type": "string",
          },
        },
        "required": [
          "message",
          "input",
          "errors",
        ],
        "type": "object",
      },
      "Person": {
        "additionalProperties": false,
        "properties": {
          "age": {
            "type": "number",
          },
          "name": {
            "type": "string",
          },
          "surname": {
            "type": "string",
          },
        },
        "required": [
          "name",
          "surname",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "title": "WunderGraph",
    "version": "0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/GetRootObj": {
      "get": {
        "operationId": "GetRootObj",
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {},
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/InvalidInputError",
                },
              },
            },
            "description": "Invalid input",
          },
        },
        "x-wundergraph-operation-type": "query",
        "x-wundergraph-requires-authentication": false,
      },
    },
    "/GetRootObj2": {
      "get": {
        "operationId": "GetRootObj2",
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {},
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/InvalidInputError",
                },
              },
            },
            "description": "Invalid input",
          },
        },
        "x-wundergraph-operation-type": "query",
        "x-wundergraph-requires-authentication": false,
      },
    },
    "/users/get": {
      "get": {
        "operationId": "GetUser",
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {},
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/InvalidInputError",
                },
              },
            },
            "description": "Invalid input",
          },
        },
        "x-wundergraph-operation-type": "query",
        "x-wundergraph-requires-authentication": false,
      },
    },
    "/users/put": {
      "post": {
        "operationId": "PutUser",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "data": {
                    "$ref": "#/components/schemas/Person",
                  },
                },
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {},
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/InvalidInputError",
                },
              },
            },
            "description": "Invalid input",
          },
        },
        "x-wundergraph-operation-type": "mutation",
        "x-wundergraph-requires-authentication": true,
      },
    },
    "/users/subscribe": {
      "get": {
        "operationId": "SubscribeToUser",
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {},
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/InvalidInputError",
                },
              },
            },
            "description": "Invalid input",
          },
        },
        "x-wundergraph-operation-type": "subscription",
        "x-wundergraph-requires-authentication": false,
      },
    },
  },
  "servers": [
    {
      "url": "http://localhost:9991/operations",
    },
  ],
}
`;
